q/tree.go

```diff
@@ -0,0 +1,89 @@
+package q
+
+import (
+	"go/token"
+	"reflect"
+)
+
+// A Criteria is used to test against a record to see if it matches.
+// It can be a combination of multiple other criterias.
+type Criteria interface {
+	// Exec is used to test the criteria against a record
+	Exec(interface{}) bool
+	exec(*reflect.Value) bool
+}
+
+type cmp struct {
+	field string
+	value interface{}
+	token token.Token
+}
+
+func (c *cmp) Exec(i interface{}) bool {
+	v := reflect.Indirect(reflect.ValueOf(i))
+	return c.exec(&v)
+}
+
+func (c *cmp) exec(v *reflect.Value) bool {
+	field := v.FieldByName(c.field)
+	return compare(field.Interface(), c.value, c.token)
+}
+
+type or struct {
+	children []Criteria
+}
+
+func (c *or) Exec(i interface{}) bool {
+	v := reflect.Indirect(reflect.ValueOf(i))
+	return c.exec(&v)
+}
+
+func (c *or) exec(v *reflect.Value) bool {
+	for _, criteria := range c.children {
+		if criteria.exec(v) {
+			return true
+		}
+	}
+
+	return false
+}
+
+type and struct {
+	children []Criteria
+}
+
+func (c *and) Exec(i interface{}) bool {
```

Comment:
@bep No but they implement the exported `Criteria` interface so people can create their own criterias.
I should improve this interface though because as it is, it is not possible to create custom criterias because of the unexported `exec` method